src/System.Memory/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -191,109 +247,215 @@ public static int GetMaxDecodedFromUtf8Length(int length)
+                else if (t2 != EncodingPad)
+                {
+                    int i2 = decodingMap[t2];
 
-                if (i0 < 0)
-                    goto InvalidExit;
-                WriteThreeLowOrderBytes(ref Unsafe.Add(ref bufferBytes, destIndex), i0);
-                destIndex += 3;
+                    i2 <<= 6;
+
+                    i0 |= i2;
+
+                    if (i0 < 0)
+                        goto InvalidExit;
+
+                    bufferBytes[destIndex] = (byte)(i0 >> 16);
+                    bufferBytes[destIndex + 1] = (byte)(i0 >> 8);
+                    destIndex += 2;
+                }
+                else
+                {
+                    if (i0 < 0)
+                        goto InvalidExit;
+
+                    bufferBytes[destIndex] = (byte)(i0 >> 16);
+                    destIndex += 1;
+                }
+
+            DoneExit:
+                bytesWritten = (int)destIndex;
+                return OperationStatus.Done;
+
+            InvalidExit:
+                bytesWritten = (int)destIndex;
+                return OperationStatus.InvalidData;
             }
-            else if (i2 != EncodingPad)
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe void Avx2Decode(ref byte* srcBytes, ref byte* destBytes, byte* srcEnd, int sourceLength, int destLength, byte* srcStart, byte* destStart)
+        {
+            // The JIT won't hoist these "constants", so help it
+            Vector256<sbyte> lutHi = s_avxDecodeLutHi;
+            Vector256<sbyte> lutLo = s_avxDecodeLutLo;
+            Vector256<sbyte> lutShift = s_avxDecodeLutShift;
+            Vector256<sbyte> mask2F = s_avxDecodeMask2F;
+            Vector256<sbyte> shuffleConstant0 = Vector256.Create(0x01400140).AsSByte();
+            Vector256<short> shuffleConstant1 = Vector256.Create(0x00011000).AsInt16();
+            Vector256<sbyte> shuffleVec = s_avxDecodeShuffleVec;
+            Vector256<int> permuteVec = s_avxDecodePermuteVec;
+
+            byte* src = srcBytes;
+            byte* dest = destBytes;
+
+            //while (remaining >= 45)
+            do
             {
-                i2 = Unsafe.Add(ref decodingMap, i2);
+                AssertRead<Vector256<sbyte>>(src, srcStart, sourceLength);
+                Vector256<sbyte> str = Avx.LoadVector256(src).AsSByte();
 
-                i2 <<= 6;
+                Vector256<sbyte> hiNibbles = Avx2.And(Avx2.ShiftRightLogical(str.AsInt32(), 4).AsSByte(), mask2F);
+                Vector256<sbyte> loNibbles = Avx2.And(str, mask2F);
+                Vector256<sbyte> hi = Avx2.Shuffle(lutHi, hiNibbles);
+                Vector256<sbyte> lo = Avx2.Shuffle(lutLo, loNibbles);
+                Vector256<sbyte> zero = Vector256<sbyte>.Zero;
 
-                i0 |= i2;
+                // https://github.com/dotnet/coreclr/issues/21247
+                if (Avx2.MoveMask(Avx2.CompareGreaterThan(Avx2.And(lo, hi), zero)) != 0)
```

Comment:
https://github.com/dotnet/coreclr/issues/21247 is on the way. I'd like to wait for that, then verify if `TestZ` produces better results.

Reply:
I think that `MoveMask` is still better since `TestZ` cannot be macro fused with the following jump but `cmp` can.

Reply:
https://github.com/dotnet/coreclr/pull/21432#discussion_r239946131

Reply:
@fiigii the version with `TestZ` would require fewer instructions, as with `MoveMask` there is

1. `Xor` (for zero-vector)
1. `And`
1. `CompareGreaterThan`
1. `MoveMask`
1. `cmp`

In contrast to just `!TestZ`. So the advantage of macro-fusion for `MoveMask` will be vanished.

Last time I checked the version with `MoveMask` showed faster results in the benchmark, though.

Reply:
Is it still a perf win if you use `TestZ` right now? If so, it might be worth just using it rather than needing to remember to come back and handle all these cases later....

Reply:
Will try to see a tendency at benchmarks.

Reply:
A version with `TestZ` is ~1...3% slower than the `MoveMask` version for input-sizes > 1000.
Same perf for smaller inputs, and didn't test larger ones.

A dasm-diff from `MoveMask` to `TestZ` shows:
```diff
G_M25171_IG06:
       C4417E6F06           vmovdqu  ymm9, ymmword ptr[r14]
       C4C13572D004         vpsrld   ymm10, ymm9, 4
       C535DBCB             vpand    ymm10, ymm10, ymm3
       C53DDBD3             vpand    ymm11, ymm9, ymm3
       C4427D00D9           vpshufb  ymm12, ymm0, ymm10
       C4427500D2           vpshufb  ymm11, ymm1, ymm11
-      C44125DBDC           vpand    ymm11, ymm11, ymm12
-      C4412564D8           vpcmpgtb ymm11, ymm11, ymm8
-      C4417DD7EB           vpmovmskb r13d, ymm11
-      4585ED               test     r13d, r13d
+      C4427D17D3           vptest   ymm11, ymm12
+      410F94C5             sete     r13b
+      450FB6ED             movzx    r13, r13b
+      4585ED               test     r13d, r13d
       743D                 je       SHORT G_M25171_IG07
```

With the fix for https://github.com/dotnet/coreclr/issues/21247 `TestZ`-version should have better code-gen, so I'll go with this options, as the improvment is then for free.

Reply:
1-3% slower than `MoveMask` still looks like a significant win as compared to the original code. I would consider just using that code now and getting the perf back when https://github.com/dotnet/coreclr/issues/21247 is resolved.

This ensures we don't need yet another work item tracking us going back to change this code later.

Reply:
Already done with https://github.com/dotnet/corefx/pull/34529/commits/8c53689df7de952018e3926978c1c07f5967cdc4